### ISSUE TITLE
support navigatable foreign key natively

### ISSUE THREAD

**Issue by hongsudt** [informatics-isi-edu/ermrestjs#203]
#### Part 1 (Get FKR related data with the same ermrest api call)

Add logic to `Reference.read()` function to change the uri in order to get the foreign key data as well as current reference data. This doesn't affect the actual `Reference._location` object and just changes the uri internally for the http request.

The `Page._data` works exactly as it used to and `Page._linkedData` stores the data for foreign keys.

#### Part 2 (retrieve visible columns and FKRs)
The current `_contextualize` function in `Columns` doesn't support outbound foreign keys. To add this support, this function should be changed as follows:
- The function should return a list of `PseudoColumn` and `Column`s.
- If the visible-columns annotation was present,
 - Check that FKRs and columns are part of table
 - Don't add repetitive values.
 - Since in some parts of reference.js we change the list of visible columns (for example in case of association table in related reference), make sure the column that you are adding is visible.
- else (heuristics) go through all the columns and:
 - If the column is not part of any FKRs, add the column.
 - If column is member of a single-column (simple) FKR: Create a `PseudoColumn` with the following displayname, and add it to the list:
     - `displayname` = FKR's `displayname`; otherwise `Column`'s displayname disambiguated with referenced `Table`'s displayname.
 - If column is member of multi-column FKR: Add the column, and create a `PseudoColumn` with the following displayname. Add the `PseudoColumn` to the end of the list.
     - `displayname` = FKR's `displayname`; otherwise referenced `Table`'s displayname. Should be disambiguated when there are multiple composite FKRs to the same table.

#### Part 3 (Inserting the values in Tuple.values) 

Currently in the `Tuple.values` we go through all the columns and change the values accordingly. Since the data for `PseudoColumn`s comes from `_linkedData` we need to differentiate between these two cases. This part can be changed,

```javascript
for (i = 0; i < this._pageRef.columns.length; i++) {
    var tempCol = this._pageRef.columns[i];
    if (tempCol._isPseudoColumn ) {
     // TODO create a url
   } else {
     // same as before
   }
}
```

For creating the url, use the following heuristics:
- Find the caption using the logic that is being used for `Tuple.displayname` (using `row-name` annotation otherwise apply the heuristics).
- For the URL: contextualize the reference to `detailed` and use the `appLink` that is defined in reference.

----------------------------
<details>

<summary> Old Descriptions </summary> 

@hongsudt commented on [Wed Apr 20 2016](https://github.com/informatics-isi-edu/chaise/issues/297)

When a column of currently displayed table is a key to another table, the record app should automatically create a url that is linked to another record page of the referenced table. 

The table annotation of the referenced table can be used to specify the title column to be used as a reference text in the display.  

See the diagram below: 
While displaying content of a construct, under the "target" column, a reference to the target record page of id=x should be displayed.
![img_20160420_155020](https://cloud.githubusercontent.com/assets/414732/14693132/28285672-0711-11e6-8382-de794fd8efda.jpg)
  

---

@robes commented on [Wed Apr 20 2016](https://github.com/informatics-isi-edu/chaise/issues/297#issuecomment-212677686)

@hongsudt: What behavior should we have when that foreign key is to a vocabulary or domain table? Do we want to drill down to see a record page for a vocabulary term. So far we have this behavior from the record page (take facebase for instance) from links to vocabulary. No one seems to get it when they go to those terms. It just confuses them. In those cases, we think that it makes more sense to go back to the search (or would be recordset) app and show records filtered by that term. In this case, when we are already on a recordset page, what should happen when the user clicks a link that is to a vocabulary term? Should it take them to the record page for vocabulary, or somehow modify the current filters, or something else?

---

@hongsudt commented on [Wed Apr 20 2016](https://github.com/informatics-isi-edu/chaise/issues/297#issuecomment-212681199)

That's really a good question. In the case of vocabulary, I think it should just display the title value as static text.  @robes I am not sure what you meant by the domain table, but supporting the case that provide a url to the search page using the title as a keyword search also makes sense. 

Is there a way to annotate the foreign key column to indicate which of the three options to use?   

---

@karlcz commented on [Wed Apr 20 2016](https://github.com/informatics-isi-edu/chaise/issues/297#issuecomment-212692584)

I think we'll need a second level of heuristics (for default) and
annotations (for override) to recognize and navigate vocabularies and
association tables differently from plain referenced entities.  Let's
think about this holistically...

A problem with vocabularies is that you can think of terms as hubs in
a hub-and-spoke reference pattern. Without domain knowledge, I'm not
sure we'd know which other referencing table to search by term if you
click on a term.  It might not always be "stay on same referencing
table" but instead might have some navigation from an obscure table to
a more canonical table, or a context menu to choose which path to
search, or even a polymorphic search results page...?

---

@carlkesselman commented on [Thu Apr 21 2016](https://github.com/informatics-isi-edu/chaise/issues/297#issuecomment-212962488)

I agree that we need to treat vocabulaires specially.  It is essential that users interactions with terms/vocabuaries be as streamlined as possible.

I have no idea what you are trying to say in the second paragraph.

Carl

---

@carlkesselman commented on [Thu Apr 21 2016](https://github.com/informatics-isi-edu/chaise/issues/297#issuecomment-212991180)

Let me add (perhaps we are in agreement), that I have long felt that we should be prescriptive about the structure of a vocabulary.  We should have standard (extensible) table structure, and we should be ok with having our tooling know about that structure.  For example, we are going to have to build UI elements for vocabulary selection that know about the structure of a term.

Carl

---

@robes commented on [Mon Apr 25 2016](https://github.com/informatics-isi-edu/chaise/issues/297#issuecomment-214475461)

@shinyichen : i think this covers one of the issues we discussed. I would suggest breaking it up as follows:

1. create a link to a record simply based on the foreign key field and value
2. then add support for using the name or "title" as the text in the anchor tag (as described above)

---

@karlcz commented on [Tue Jun 21 2016](https://github.com/informatics-isi-edu/chaise/issues/297#issuecomment-227568130)

Should this issue be retargeted to record2 app?

</details>

**Comment by hongsudt** [comment]
* create a getter/mapper that maps between columns and foreign-key references that it is a part of. This will help determine whether a column is involved in one FKR or more than one FKR. 
* show/hide foreign-keys (this can be done prior to the read call): 
  - if 2016 visible column is defined, follow the annotation, else 
  - single-key FKR: go through each column defined in the Reference.columns, if the column involves in FKRs, replace the column with those FKRs.
  - composite-key FKR:
      - leave the participating columns as is, and insert the pseudo-columns in the order in which the FKRs are returned.   
      - make sure that the each FKR is displayed only once
      - if one or more of FKR's participating columns are not visible, FKR is not visible. 
* Pseudo-column names:
  - FKR's to_name if available
  - single-key FKR: use the column's displayName disambiguated with the referenced table's displayName    
  - composite-key FKR: use the referenced table's displayName?
* pseudo-column values (inserted in Tuple.values): 
  - caption: is defined by row_name context of 2016 Table Display
  - href: a url generated by using an app_link and Reference.location object. Should this be a utility function?  Note: if the row_name context already define a navigational link, don't override it. 

@robes @karlcz any comments? I incorporated the text from ermrestjs/doc/heuristics.md in this note as well.

**Comment by karlcz** [comment]
The annotation spec for Visible Columns allows bare column names `"Column X"` and foreign key names `["Schema X", "Constraint Y"]` to both appear in the list.  This allows the DBA to make participating columns and/or keys visible or invisible in any combination and any order.

I would implement the default heuristic as a preprocessing step to create an effective Visible Columns list if one was not provided by annotation.  Then, I would make the main code always process this list regardless of which way it was obtained.

**Comment by robes** [comment]
@hongsudt So if I follow the main differences between your sketch and

https://github.com/informatics-isi-edu/ermrestjs/blob/master/doc/heuristics.md#outbound-references

are some more specificity about the column name and displayName and the introduction of this new app_link annotation?

It seems fine to me.

* This sentence in your text "leave the participating columns as is, and insert the pseudo-columns in the order in which the FKRs are returned" should probably replace "insert" with "append" because in this case they will just append to the end of the columns.
* "composite-key FKR: use the referenced table's displayName?": we will need some way to disambiguate.

**Comment by RFSH** [comment]
@hongsudt @robes @karlcz 

I update the issue description, can you please take a look?

These are some parts that I am not sure about,

1. Should I create a new class for pseudo-columns? or just add a `isPseudoColumn` to Column? What should be its attributes?
2. For the URL:
 - If the `row_name` context wasn't available, what should I use for the caption?
 - Can you elaborate more on href part? Should I use [2016 App Links](https://github.com/informatics-isi-edu/ermrest/blob/master/user-doc/annotation.md#2016-app-links) for that?

**Comment by robes** [comment]
First, comments on the new task description.

Part 1 of your outline:
- Looks like the `_contextualize` function is in `Columns` not `Table`
- I would think `Columns._contextualize` should just return `Column`s and `PseudoColumn`s
- returning `ForeignKeyRef`s mixed with `Column`s doesn't seem ideal

Part 2:
- regarding the `name` and `displayName` please refer to https://github.com/informatics-isi-edu/ermrestjs/blob/master/doc/heuristics.md#pseudo-column-heuristics
- yes, the pseudo-column should also have a `type` property. Seems like that will probably be `'markdown'` unless I'm mistaken
- it can also have `comment` which can be taken from the `ForeignKeyRef.comment`
- It should have all the same _public_ properties as the `Column`. Pick sensible defaults if they cannot be drawn from the `ForeignKeyRef` (or `Column` for single-column FKRs)

Part 3:
- No, I don't think `tuples.values` needs to change or needs a separate conditional code path, unless I'm mistaken. You just need a different implementation for the  `PseudoColumn.formatPresentation` function.

Now, to your last comment:
1. Yes, try making a new `PseudoColumn` class for this, as a subclass of `Column` (or at least by convention by cloning the parent class's prototype) and then overriding methods with different implementations for the pseuo-column. It should have all the same public attributes that Column has.
2. For the URL
  - Yes, probably use the displayname then
  - Right, seems like you need to use the App Link annotation for the `href`
